src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -1,197 +1,599 @@
+
+            public abstract float NextSingle();
+
+            public abstract void NextBytes(byte[] buffer);
+
+            public abstract void NextBytes(Span<byte> buffer);
+        }
+
+        #region Implementations
+        /// <summary>
+        /// Provides an implementation of the xoshiro256** algorithm. This implementation is used
+        /// when no seed is specified and an instance of the base Randomclass is constructed.  As
+        /// such, we are free to implement however we see fit, withoutback compat concerns around
+        /// the sequence of numbers generated or what methods call what other methods.
+        /// </summary>
+        private sealed class Xoshiro256StarStarImpl : ImplBase
         {
-            if (buffer is null)
+            // NextUInt64 is based on the algorithm from http://prng.di.unimi.it/xoshiro256starstar.c:
+            //
+            //     Written in 2018 by David Blackman and Sebastiano Vigna (vigna@acm.org)
+            //
+            //     To the extent possible under law, the author has dedicated all copyright
+            //     and related and neighboring rights to this software to the public domain
+            //     worldwide. This software is distributed without any warranty.
+            //
+            //     See <http://creativecommons.org/publicdomain/zero/1.0/>.
+
+            private ulong _s0, _s1, _s2, _s3;
+
+            public unsafe Xoshiro256StarStarImpl()
             {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.buffer);
+                do
+                {
+                    Guid g;
+
+                    // Seed the four ulongs (4 64-bit values) from two Guids (2 128-bit values)
+                    g = Guid.NewGuid();
+                    _s0 = *(ulong*)&g;
+                    _s1 = *((ulong*)&g + 1);
+
+                    g = Guid.NewGuid();
+                    _s2 = *(ulong*)&g;
+                    _s3 = *((ulong*)&g + 1);
+                }
+                while ((_s0 | _s1 | _s2 | _s3) == 0); // at least one value must be non-zero
             }
 
-            for (int i = 0; i < buffer.Length; i++)
+            /// <summary>Produces a value in the range [0, ulong.MaxValue].</summary>
+            private ulong NextUInt64()
             {
-                buffer[i] = (byte)InternalSample();
+                ulong result = BitOperations.RotateLeft(_s1 * 5, 7) * 9;
+                ulong t = _s1 << 17;
+
+                _s2 ^= _s0;
+                _s3 ^= _s1;
+                _s1 ^= _s2;
+                _s0 ^= _s3;
+
+                _s2 ^= t;
+                _s3 = BitOperations.RotateLeft(_s3, 45);
+
+                return result;
             }
-        }
 
-        public virtual void NextBytes(Span<byte> buffer)
-        {
-            for (int i = 0; i < buffer.Length; i++)
+            public override int Next()
             {
-                buffer[i] = (byte)Next();
+                while (true)
```

Comment:
Are there any concerns about this running for "many loops"?

It probably won't happen in practice, but since this is "random"....

Reply:
If it loops three times or even twice ever in the use of this product we have a bigger (randomness) problem presumably?

Reply:
That's the "problem" with random, you never know if it's a bug or just randomness 😄 

(If it's "truly random", you are just as likely to get `0` 5x in a row as `5` unique values 😉)

Reply:
Right, I made no claim about randomness, just that if this was not "beyond-beyond-vanishingly unlikely to be an issue" then we may have a flaw in the generator, rather than a perf problem.

Reply:
This is the rejection sampling loop which produces a uniform distributiion in the interval [0, maxValue) with no bias whatsoever. Whereas the previous approach relied on floating point multiplication, and therefore I think probably exhibited some bias; certainly I would say that use of floating point arithmetic makes it harder to reason about whether it is biased or not.

In the worst case scenario maxValue is 1 above a power of 2 (2^n +1), and thus each loop has an approx. 50% chance of generating a value that can be returned. E.g. maxValue = 129, and bits = 8 giving 2^8=256.  

If you were to call this method a lot, say 100s of millions or billions of times (which is not unlikely/unreasonable) with such a worst case maxValue, then you can expect the following behaviour on average:

1 loop 50% of calls
2 loops 25% of calls
3 loops 12.5% of calls,
etc.

So roughly speaking: 
  ` percentage of calls = 1/(2^loops)`

Re-arranging to give:
   `loops = log2(1/p)`

So in a billion calls (for example) we can expect one of those calls (on average) to run for this many loops:
  `loops = log2(10^9) = 29`

Let's just sanity check that number. In a billion calls we can expect a 1 in a billion probability event to occur (on average, if we did lots of experiments, doing 1 billion calls in each experiment). One such '1 in a billion' event would be executing the loop 30 times (or thereabouts), with a 50% chance of succeeding on each loop!!

So 30 loops is a tiny fraction of overall executions, but not an uncommon occurrence (it could perhaps occur once every few seconds on a fast CPU calling this method heavily).

However, let's just say we put a hard limit of 100 loops (I pulled that number out of the air, but let's just run with that for now) after which we exit to prevent any chance of a bad PRNG causing an infinite loop (or lots more loops on average than we'd like). We probably shouldn't throw an exception, because it might occur on rare occasions - OK maybe at  a limit of 100 loops the chance is somewhat theoretical, but it doesn't feel like a good pattern/precedent I would say.

So the other option is to just pick a 'close enough' value in that very rare (maybe once in lifetime of the universe!) scenario. Well, we can take the sample we have that is too high, and divide by two (shift right), or we can zero out the high bit to give a sample within the required interval. These special case samples are not properly distributed, but they are so rare that the overall bias of Next(int) is vanishingly small.

Hence my point is that there is a way of putting in a limit to the number of loops as a safety check, and that doesn't break the overall contract of that method in any meaningful way. What would be a sane limit? well, greater than 30, as we have seen that that is a common occurrence. If we said a billion times less likely than the 30 loop case we get to 60 loops:

`log2(10^18) = 59.79 `to be precise.

I'm not suggesting this is the way forward, but thought some might find it a useful thought exercise :)